modules/core/csv/src/Encoder/CsvEncoder.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace Drupal\farm_csv\Encoder;
+
+use Drupal\csv_serialization\Encoder\CsvEncoder as ContribCsvEncoder;
+
+/**
+ * Extend the contrib CsvEncoder class with custom header extraction logic.
+ */
+class CsvEncoder extends ContribCsvEncoder {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function extractHeaders(array $data, array $context = []) {
+    $headers = [];
+
+    // Iterate over each row and merge its headers.
+    // This differs from the parent method in that it combines headers from all
+    // rows instead of only using headers from the first row.
+    foreach ($data as $row) {
+      $headers = array_merge($headers, array_diff(array_keys($row), $headers));
+    }
```

Comment:
As discussed on the dev call, using a mutable structure for accumulating the header keys should be much more efficient. It turns out the solution can be simpler than I thought too since PHP conserves key ordering in associative arrays. https://www.php.net/manual/en/language.types.array.php

```suggestion
  // Iterate over each row and accumulate the unique headers.
  // This differs from the parent method in that it combines headers from all
  // rows instead of only using headers from the first row.
  // This relies on the property of PHP arrays being ordered maps
  // so the header keys will appear in the order they are first encountered
  // when iterating over all rows.
  $unique_headers = [];
  foreach ($data as $row) {
    foreach (array_keys($row) as $header) {
      if (!array_key_exists($header, $unique_headers)) {
        // The value assigned doesn't matter here
        $unique_headers[$header] = 1;
      }
    }
  }

  $headers = array_keys($unique_headers);
```